### PR TITLE
Bug 1354939 - Prevent NSKeyedUnarchiver from not throwing by setting the decodingFailurePolicy.

### PR DESCRIPTION
**Pull request by farhanpatel**
The issues with restoring tabs was, when decoding a child object - if the child object failed to decode NSCoder would throw an exception. There was no way to prevent this. But there is! Since iOS9 at least. 

There is a flag `decodingFailurePolicy` which needs to be set to `setErrorAndReturn` and this policy will be applied to all children of the object being decoded. 

This could be a risky change. This assumes that the only crashes the try/catch blocks were catching were the above cases before. (And the fix only handles restoring related crashes). 

How can we actually make the preserving (encoding) part crash? We do have some vague crashes related to dealloc. But that could be related to the Try/catch block. Hard to say.

### CODE REVIEW

**Comment by farhanpatel** [comment]
### SwiftLint found issues

#### Warnings

File | Line | Reason |
| --- | ----- | ----- |
TabManagerTests.swift | 383 | Limit vertical whitespace to a single empty line. Currently 2. 


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: Client/Frontend/Browser/TabManager.swift [hunk 1]
```diff
@@ -670,16 +672,12 @@ extension TabManager {
         let archiver = NSKeyedArchiver(forWritingWith: tabStateData)
         archiver.encode(savedTabs, forKey: "tabs")
         archiver.finishEncoding()
-        tabStateData.write(toFile: path, atomically: true)
+        let status = tabStateData.write(toFile: path, atomically: true)
+        print(status)
```
Comment (sleroux) [inline Client/Frontend/Browser/TabManager.swift]: Nit: Move to logger if we want to keep the print around?

File: ClientTests/TabManagerTests.swift [hunk 2]
```diff
@@ -335,4 +335,50 @@ class TabManagerTests: XCTestCase {
...
+    func testFailedRestore() {
+        // Lets use a bad tab tate
+        let filePath =  Bundle(for: type(of: self)).path(forResource: "badTabsState", ofType: "archive")!
+        let file = try! Data(contentsOf: URL(fileURLWithPath: filePath))
+        try! file.write(to: URL(fileURLWithPath: TabManager.tabsStateArchivePath()))
+
+        // lets create a Tabmanager and try a restore
+        let profile = TabManagerMockProfile()
+        let manager = TabManager(prefs: profile.prefs, imageStore: nil)
+        manager.restoreTabs()
```
Comment (sleroux) [inline ClientTests/TabManagerTests.swift]: > How can we actually make the preserving (encoding) part crash? We do have some vague crashes related to dealloc. But that could be related to the Try/catch block. Hard to say.

Is this what you were referring to? Finding a way to fail to restore so we can assert some kind of failure condition here? You could try emulating the 7.x tab loss case by creating an archive with a different namespace'd class and try pulling it back out with the current one.

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 3]
```diff
@@ -530,7 +530,7 @@ class BrowserViewController: UIViewController {
         if PLCrashReporter.shared().hasPendingCrashReport() {
             PLCrashReporter.shared().purgePendingCrashReport()
             showRestoreTabsAlert()
-        } else {
+        } else if !AppConstants.IsRunningTest {
```
Comment (st3fan) [inline Client/Frontend/Browser/BrowserViewController.swift]: Don't we have tests that depend on tabs being restored between sessions?

File: Client/Frontend/Browser/TabManager.swift [hunk 4]
```diff
@@ -670,16 +672,12 @@ extension TabManager {
...
     }
 
     func preserveTabs() {
-        // This is wrapped in an Objective-C @try/@catch handler because NSKeyedArchiver may throw exceptions which Swift cannot handle
```
Comment (st3fan) [inline Client/Frontend/Browser/TabManager.swift]: Why don't we keep this? I think the `NSKeyedArchiver` also still throws exceptions, right? Did we not see stack traces where it threw exceptions? Better to wrap it in a handler then?

File: Client/Frontend/Browser/BrowserViewController.swift (hunk 3 above)
Comment (farhanpatel) [inline Client/Frontend/Browser/BrowserViewController.swift]: we do. I didnt add this. This was initially in the `restoretabs()` in TabManager section but I didn't want that behavior there as the test would never restore tabs. So I moved it up into BVC in order to keep the behaviour for tests that need it.

File: Client/Frontend/Browser/TabManager.swift (hunk 4 above)
Comment (farhanpatel) [inline Client/Frontend/Browser/TabManager.swift]: Yea. Thats what I was trying to understand. How could NSKeyedArchiver actually crash? The one you pointed me to was a weird dealloc case. That seems like something that could be caused by our try/catch hack? How often did that crash happen?

**Pull request closed without merging**